extmod/vfs_fat.c

```diff
@@ -111,22 +113,127 @@ STATIC mp_obj_t fat_vfs_mkdir(mp_obj_t vfs_in, mp_obj_t path_o) {
+    }
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_2(fat_vfs_chdir_obj, fat_vfs_chdir);
+
+// Checks for path equality, ignoring trailing slashes:
+//   path_equal(/, /) -> true
+// second argument must be in canonical form (meaning no trailing slash, unless it's just /)
+STATIC bool path_equal(const char *path, const char *path_canonical) {
+    while (*path_canonical != '\0' && *path == *path_canonical) {
+        ++path;
+        ++path_canonical;
+    }
+    if (*path_canonical != '\0') {
+        return false;
+    }
+    while (*path == '/') {
+        ++path;
+    }
+    return *path == '\0';
+}
+
+/// \function stat(path)
+/// Get the status of a file or directory.
+STATIC mp_obj_t fat_vfs_stat(mp_obj_t vfs_in, mp_obj_t path_in) {
+    (void)vfs_in;
+    const char *path = mp_obj_str_get_str(path_in);
+
+    FILINFO fno;
+#if _USE_LFN
+    fno.lfname = NULL;
+    fno.lfsize = 0;
+#endif
+
+    FRESULT res;
+    if (path_equal(path, "/")) {
+        // stat root directory
+	fno.fsize = 0;
+	fno.fdate = 0;
+	fno.ftime = 0;
+	fno.fattrib = AM_DIR;
+    } else {
+        res = FR_NO_PATH;
+        for (size_t i = 0; i < MP_ARRAY_SIZE(MP_STATE_PORT(fs_user_mount)); ++i) {
+            fs_user_mount_t *vfs = MP_STATE_PORT(fs_user_mount)[i];
+            if (vfs != NULL && path_equal(path, vfs->str)) {
+                // stat mounted device directory
+                fno.fsize = 0;
+                fno.fdate = 0;
+                fno.ftime = 0;
+                fno.fattrib = AM_DIR;
+                res = FR_OK;
+            }
+        }
+        if (res == FR_NO_PATH) {
+            // stat normal file
+            res = f_stat(path, &fno);
+        }
+        if (res != FR_OK) {
+            nlr_raise(mp_obj_new_exception_arg1(&mp_type_OSError,
+                MP_OBJ_NEW_SMALL_INT(fresult_to_errno_table[res])));
+        }
+    }
+
+    mp_obj_tuple_t *t = MP_OBJ_TO_PTR(mp_obj_new_tuple(10, NULL));
+    mp_int_t mode = 0;
+    if (fno.fattrib & AM_DIR) {
+        mode |= 0x4000; // stat.S_IFDIR
+    } else {
+        mode |= 0x8000; // stat.S_IFREG
+    }
+    mp_int_t seconds = timeutils_seconds_since_2000(
+        1980 + ((fno.fdate >> 9) & 0x7f),
+        (fno.fdate >> 5) & 0x0f,
+        fno.fdate & 0x1f,
+        (fno.ftime >> 11) & 0x1f,
+        (fno.ftime >> 5) & 0x3f,
+        2 * (fno.ftime & 0x1f)
+    );
+    t->items[0] = MP_OBJ_NEW_SMALL_INT(mode); // st_mode
+    t->items[1] = MP_OBJ_NEW_SMALL_INT(0); // st_ino
+    t->items[2] = MP_OBJ_NEW_SMALL_INT(0); // st_dev
+    t->items[3] = MP_OBJ_NEW_SMALL_INT(0); // st_nlink
+    t->items[4] = MP_OBJ_NEW_SMALL_INT(0); // st_uid
+    t->items[5] = MP_OBJ_NEW_SMALL_INT(0); // st_gid
+    t->items[6] = MP_OBJ_NEW_SMALL_INT(fno.fsize); // st_size
+    t->items[7] = MP_OBJ_NEW_SMALL_INT(seconds); // st_atime
+    t->items[8] = MP_OBJ_NEW_SMALL_INT(seconds); // st_mtime
+    t->items[9] = MP_OBJ_NEW_SMALL_INT(seconds); // st_ctime
+
+    return (mp_obj_t)t;
```

Comment:
OK. I'll start with preparation - replacing the text error messages in
vfs_fat.c with the Posix number messages.

On 27.05.2016 21:22, Paul Sokolovsky wrote:
> In extmod/vfs_fat.c
> <https://github.com/micropython/micropython/pull/2110#discussion_r64955190>:
> 
>> +        (fno.ftime >> 11) & 0x1f,
>> +        (fno.ftime >> 5) & 0x3f,
>> +        2 * (fno.ftime & 0x1f)
>> +    );
>> +    t->items[0] = MP_OBJ_NEW_SMALL_INT(mode); // st_mode
>> +    t->items[1] = MP_OBJ_NEW_SMALL_INT(0); // st_ino
>> +    t->items[2] = MP_OBJ_NEW_SMALL_INT(0); // st_dev
>> +    t->items[3] = MP_OBJ_NEW_SMALL_INT(0); // st_nlink
>> +    t->items[4] = MP_OBJ_NEW_SMALL_INT(0); // st_uid
>> +    t->items[5] = MP_OBJ_NEW_SMALL_INT(0); // st_gid
>> +    t->items[6] = MP_OBJ_NEW_SMALL_INT(fno.fsize); // st_size
>> +    t->items[7] = MP_OBJ_NEW_SMALL_INT(seconds); // st_atime
>> +    t->items[8] = MP_OBJ_NEW_SMALL_INT(seconds); // st_mtime
>> +    t->items[9] = MP_OBJ_NEW_SMALL_INT(seconds); // st_ctime
>> +
>> +    return (mp_obj_t)t;
> 
> there's also c): please-please, don't try to make several changes at
> once! "stat(), chdir() and rmdir()", should be (at least) 3 separate
> pull requests. No need to start with next one until previous is
> finished, submitted, ran thru CI, reviewed, merged. @robert-hh
> <https://github.com/robert-hh> , your previous PR was the right way to
> do it. It's better to spend a day on each, than a week (or more) on
> something which tries to pack them together.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/micropython/micropython/pull/2110/files/172a3fb51f31f4789e385828c57326417ecc8cdc#r64955190>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AL5hxIaJZcn-n4IfnjAoqUHuKjRdh_bwks5qF0R_gaJpZM4Imtsp>.
>